### PR TITLE
analytical-patterns homework

### PR DESCRIPTION
**Pull request by nlubalo**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly implemented the state change tracking for NBA players, and your code covers all aspects of the prompt. However, there are a few areas where your solution could be improved. 

1) **Readability and Comments:** Your code is generally well-commented, which is good practice. However, the comments at the beginning of your code are commented out, which makes it a bit confusing. If these are instructions or notes to yourself, consider making them more clear or removing them if they are not necessary.

2) **Error Handling:** Your code includes an 'ERROR' state in the case statement, which is a good start for error handling. However, it would be helpful to provide more specific error messages to make it easier to debug if something goes wrong. For example, you could specify what kind of error occurred, such as 'Invalid player state' or 'Invalid season'.

3) **Modularity and Reusability:** Your code is modular and well-structured, which makes it easy to understand and reuse specific components. However, you could improve the reusability of your code by defining some parts of your query as reusable functions or procedures. For example, the logic for determining the player's state could be defined as a separate function.

4) **Robustness:** Your code seems to handle the main cases well, but it's not clear how it would handle edge cases. For example, what happens if a player's name is null, or if the seasons are not consecutive? It would be good to add some checks or validations for these cases.

5) **Variable Names:** Your variable names are generally meaningful and descriptive, which makes your code easier to read. However, the variable name 'is_active' could be more descriptive. Consider renaming it to something like 'is_player_active' to make it clear that it refers to the player's active status.

In conclusion, your solution is quite good, but there are a few areas where it could be improved. Keep up the good work, and keep these suggestions in mind for your future assignments.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_1.sql [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+---CREATE OR REPLACE TABLE nancyatienno21998.nba_players_tracker (
+---    player_name VARCHAR,
+--    first_active_season INT,
+---    last_active_season INT,
+---    seasons_active ARRAY(INT),
+---    season_active_state VARCHAR,
+---    season INT
+---    ) WITH
+--    (format = 'PARQUET',
+---    partitioning = ARRAY['season'])
```
Comment (HjersmanJ) [inline submission/query_1.sql:17]: This can be left uncommented.

File: submission/query_1.sql [hunk 2]
```diff
@@ -0,0 +1,64 @@
+--Write a query (query_1) that does state change tracking for nba_players. Create a state change-tracking field that takes on the following values:
+--A player entering the league should be New
+--A player leaving the league should be Retired
+--A player staying in the league should be Continued Playing
+--A player that comes out of retirement should be Returned from Retirement
+--A player that stays out of the league should be Stayed Retired
```
Comment (HjersmanJ) [inline submission/query_1.sql:6]: This would be better summarized into the README file.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly implemented the state change tracking for NBA players, and your code covers all aspects of the prompt. However, there are a few areas where your solution could be improved.

1) **Readability and Comments:** Your code is generally well-commented, which is great. However, the comments at the beginning of your code are commented out, which makes it a bit confusing. If these are instructions or notes to yourself, consider making them more clear or removing them if they are not necessary.

2) **Error Handling:** Your code includes an 'ERROR' state in the case statement, which is a good start for error handling. However, it would be beneficial to provide more specific error messages to help with debugging. For example, you could specify what kind of error occurred, such as 'Invalid player state' or 'Invalid season data'.

3) **Robustness:** Your code assumes that the `is_active` field in the `current_season` table accurately reflects whether a player is active in the current season. However, it would be more robust to determine this based on whether the player has a record in the `current_season` table. This would handle cases where the `is_active` field might be incorrect.

4) **Modularity and Reusability:** Your code is well-structured and the use of CTEs (Common Table Expressions) makes it modular and easy to understand. However, the `combined` CTE is quite large and does a lot of different things. Consider breaking it up into smaller CTEs to improve readability and reusability.

Here's a suggestion on how you could improve your code:

```sql
-- Instead of using is_active to determine if a player is active, check if they have a record in the current_season table
-- This makes the code more robust to errors in the is_active field
current_season AS (
  SELECT 
    player_name,
    current_season,
    CASE WHEN player_name IS NOT NULL THEN TRUE ELSE FALSE END AS is_active
  FROM bootcamp.nba_players
  where current_season = 1996
),

-- Break up the combined CTE into smaller CTEs
player_seasons AS (
  SELECT
    COALESCE(ls.player_name, cs.player_name)     AS player_name,
    COALESCE(ls.first_active_season, cs.current_season) as first_active_season,
    COALESCE(cs.current_season, ls.last_active_season) as last_active_season,
    COALESCE(ls.season + 1, cs.current_season) AS season
  FROM last_season ls
  FULL OUTER JOIN current_season cs ON ls.player_name = cs.player_name
),
seasons_active AS (
  SELECT
    player_name,
    CASE
      WHEN ls.seasons_active IS NULL THEN ARRAY[cs.current_season]
      WHEN cs.current_season IS NULL THEN ls.seasons_active
      ELSE ls.seasons_active || ARRAY[cs.current_season]
    END AS seasons_active
  FROM player_seasons ps
  LEFT JOIN last_season ls ON ps.player_name = ls.player_name
  LEFT JOIN current_season cs ON ps.player_name = cs.player_name
)
```

Keep up the good work! Your solution is on the right track, and with a few improvements, it will be even better.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_1.sql [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  first_active_season,
+  last_active_season,
+  seasons_active,
+  CASE
+    WHEN is_active and first_active_season - last_active_season = 0 THEN 'New'
+    WHEN is_active and season - last_active_season =1 THEN 'Continued Playing'
+    WHEN is_active and season - last_active_season >1 THEN 'Returned from Retirement'
+    WHEN NOT is_active and season - last_active_season = 1 THEN 'Retired'
+    WHEN NOT is_active and season - last_active_season >1 THEN 'Stayed Retired'
+    ELSE 'ERROR'
```
Comment (HjersmanJ) [inline submission/query_1.sql:61]: If you are including an `ERROR` clause, ideally there would be somewhere showing that you checked for it.

File: submission/query_2.sql [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        WHEN GROUPING (gdd.player_name, gdd.team_abbreviation) = 0 THEN 'player_and_team'
+        WHEN GROUPING (gdd.player_name, g.season) = 0 THEN 'player_and_season'
+        WHEN GROUPING (gdd.team_abbreviation) = 0 THEN 'team'
+    END as aggregation_level,
+  -- handle cases where player_name or team_abbreviation is NULL
+  COALESCE(gdd.player_name, 'Overall') AS player_name,
+  COALESCE(gdd.team_abbreviation, 'Overall') AS team,
+  -- sum of points for the aggregation
+  SUM(gdd.pts) AS points,
+  -- handle cases where season is NULL
```
Comment (HjersmanJ) [inline submission/query_2.sql:19]: These comments are a bit unnecessary since the code makes them clear. Would recommend omitting.

**Review by HjersmanJ — APPROVED** [review]
No feedback for code accuracy since output is perfect. However, there were a few tips on documentation that could make your code more readable and therefore more maintainable. Great job!
